### ISSUE TITLE
add an implementation of __aeabi_memcpy and related memory builtins to compiler-rt

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#4189]
Normally, we can port compiler-rt functions. However, [LLVM's function is incorrect](https://lists.llvm.org/pipermail/llvm-dev/2016-April/098174.html).

This task involves understanding what the correct ABI is for these functions, and figuring out how to have high performing (or at least correct) `memcpy`, `memmove`, and `memset` functions, that adheres to the ABI.

**Comment by LemonBoy** [comment]
Please note that this is also true for the EABI counterparts of `memmove` and `memset`
.

**Comment by vegecode** [comment]
# 4.3.4  Memory copying, clearing, and setting

## Memory copying

Memcpy-like helper functions are needed to implement structure assignment. We define three functions providing various levels of service, in addition to the normal ANSI C memcpy, and three variants of memmove.

```
void __aeabi_memcpy8(void *dest, const void *src, size_t n); 
void __aeabi_memcpy4(void *dest, const void *src, size_t n); 
void __aeabi_memcpy(void *dest, const void *src, size_t n); 

void __aeabi_memmove8(void *dest, const void *src, size_t n); 
void __aeabi_memmove4(void *dest, const void *src, size_t n); 
void __aeabi_memmove(void *dest, const void *src, size_t n);
```

These functions work like the ANSI C memcpy and memmove functions. However, __aeabi_memcpy8 mayassume that both of its arguments are 8-byte aligned, __aeabi_memcpy4 that both of its arguments are 4-byte aligned. None of the three functions is required to return anything in r0.

Each of these functions can be smaller or faster than the general memcpy or each can be an alias for memcpy itself, similarly for memmove.

Compilers can replace calls to memcpy with calls to one of these functions if they can deduce that the constraints are satisfied. For example, any memcpy whose return value is ignored can be replaced with __aeabi_memcpy. If the copy is between 4-byte-aligned pointers it can be replaced with __aeabi_memcpy4, and so on.

The size_t argument does not need to be a multiple of 4 for the 4/8-byte aligned versions, which allows copies with a non-constant size to be specialized according to source and destination alignment.

Small aligned copies are likely to be inlined by compilers, so these functions should be optimized for larger copies.

## Memory clearing and setting

In similar deference to run-time efficiency we define reduced forms of memset and memclr.

```
void __aeabi_memset8(void *dest, size_t n, int c); 
void __aeabi_memset4(void *dest, size_t n, int c); 
void __aeabi_memset(void *dest, size_t n, int c); 

void __aeabi_memclr8(void *dest, size_t n); 
void __aeabi_memclr4(void *dest, size_t n); 
void __aeabi_memclr(void *dest, size_t n); 
```

Note that relative to ANSI memset, __aeabi_memset has the order of its second and third arguments reversed. This allows __aeabi_memclr to tail-call __aeabi_memset.

The memclr functions simplify a very common special case of memset, namely the one in which c = 0 and the memory is being cleared to all zeroes.

The size_t argument does not need to be a multiple of 4 for the 4/8-byte aligned versions, which allows clears and sets with a non-constant size to be specialized according to the destination alignment.

__In general__, implementations of these functions are allowed to corrupt only the integer core registers permitted to be corrupted by the [AAPCS] (r0-r3, ip, lr, and CPSR).

If there is an attached device with efficient memory copying or clearing operations (such as a DMA engine), its device supplement specifies whether it may be used in implementations of these functions and what effect such use has on the device’s state.

**Comment by vegecode** [comment]
What do you think the significance of "in general" is? I don't see that phrase anywhere else in the run time abi document.

**Comment by vegecode** [comment]
Just dropping the section off for easy finding.

Thoughts:

It would be a negative in the microcontroller world to have two versions of memcpy due to the extra binary size.

Is it possible to reuse code from the standard library but set it to using the AAPCS calling convention instead of AAPCS_VFP? I guess the standard library is preferred to use the zig calling convention for maximum compiler optimizations, though.

In what situations is the compiler producing these calls when compiling zig code? If it is only producing these calls when compiling C code and the compiler-rt library just needs to support that, then the extra binary size could just be considered the cost of doing business with C.

**Comment by greenfork** [comment]
I'd like to take this issue but I have zero knowledge about this stuff. I found:
- Run-time ABI documentation https://developer.arm.com/documentation/ihi0043/e?lang=en
- AAPCS procedure call standard documentation https://developer.arm.com/documentation/ihi0042/latest

My understanding is that I will need to write some assembly code in 
https://github.com/ziglang/zig/blob/1534cd2f88d94f320a5735b55c0ce3aebb905b6c/lib/std/special/compiler_rt/arm.zig#L28-L33

And also add something to
https://github.com/ziglang/zig/blob/1534cd2f88d94f320a5735b55c0ce3aebb905b6c/src/codegen.zig#L3466-L3482

After that I will need to devise some tests and use qemu to test it, I found notes on testing in [CONTRIBUTING](https://github.com/ziglang/zig/blob/master/CONTRIBUTING.md).

I basically want to go from zero to hero, I would appreciate any existing examples, introductory materials or keywords I need to google to complete this issue.

**Comment by matu3ba** [comment]
As of d24ebf1d12cf66665b52136a2807f97ff021d78d in lib/compiler_rt/arm.zig there is
```zig
comptime {
    if (!builtin.is_test) {
        if (arch.isARM() or arch.isThumb()) {
            @export(__aeabi_unwind_cpp_pr0, .{ .name = "__aeabi_unwind_cpp_pr0", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_unwind_cpp_pr1, .{ .name = "__aeabi_unwind_cpp_pr1", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_unwind_cpp_pr2, .{ .name = "__aeabi_unwind_cpp_pr2", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_ldivmod, .{ .name = "__aeabi_ldivmod", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_uldivmod, .{ .name = "__aeabi_uldivmod", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_idivmod, .{ .name = "__aeabi_idivmod", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_uidivmod, .{ .name = "__aeabi_uidivmod", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_memcpy, .{ .name = "__aeabi_memcpy", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memcpy4, .{ .name = "__aeabi_memcpy4", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memcpy8, .{ .name = "__aeabi_memcpy8", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_memmove, .{ .name = "__aeabi_memmove", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memmove4, .{ .name = "__aeabi_memmove4", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memmove8, .{ .name = "__aeabi_memmove8", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_memset, .{ .name = "__aeabi_memset", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memset4, .{ .name = "__aeabi_memset4", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memset8, .{ .name = "__aeabi_memset8", .linkage = common.linkage, .visibility = common.visibility });

            @export(__aeabi_memclr, .{ .name = "__aeabi_memclr", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memclr4, .{ .name = "__aeabi_memclr4", .linkage = common.linkage, .visibility = common.visibility });
            @export(__aeabi_memclr8, .{ .name = "__aeabi_memclr8", .linkage = common.linkage, .visibility = common.visibility });

            if (builtin.os.tag == .linux) {
                @export(__aeabi_read_tp, .{ .name = "__aeabi_read_tp", .linkage = common.linkage, .visibility = common.visibility });
            }
        }
    }
}
```

What is missing for this issue to be closed?

**Comment by matu3ba** [comment]
I think this issue should be closed, as all memory runtime routines for ARM seem to be implemented: #14634.
If anything is missing, we can reopen this issue.
